File: apps/api/app/schemas/discovery.py
from typing import Literal
from uuid import UUID

from pydantic import BaseModel


class SportProfileSummary(BaseModel):
    sport: str
    level: str
    gym_name: str | None = None
    golf_club: str | None = None


class PartnerCardResponse(BaseModel):
    user_id: UUID
    display_name: str
    suburb: str | None = None
    # Truncated 160-char preview for the feed card. Full text lives in `bio`.
    bio_excerpt: str | None = None
    # Full bio for the profile-detail preview (V1 partner detail modal).
    bio: str | None = None
    avatar_url: str | None = None
    # Ordered list of all profile photos (avatar_url is photo_urls[0] when set).
    # Empty when the user has not uploaded any photos.
    photo_urls: list[str] = []
    age: int | None = None
    sport_profiles: list[SportProfileSummary]


class DiscoveryFeedResponse(BaseModel):
    items: list[PartnerCardResponse]
    total: int
    limit: int
    offset: int


class RecordActionRequest(BaseModel):
    target_user_id: UUID
    action: Literal["like", "pass", "save"]
    sport: Literal["gym", "golf", "tennis", "running"]


class RecordActionResponse(BaseModel):
    action: str
    match_created: bool
    match_id: UUID | None = None

File: packages/shared-types/src/discovery.ts
/**
 * Discovery domain types.
 *
 * Discovery is the browsing surface — the Discover tab in the mobile app.
 *
 * A PartnerCard is the summary shown in the discovery feed.
 * A DiscoveryAction is a pass/like/save gesture by the current user.
 * When two users mutually like each other, a Match is created.
 */

import type { Paginated, UUID } from './common';
import type { FitnessLevel, PreferredTime, Sport } from './sport-profile';

// ---------------------------------------------------------------------------
// Actions
// ---------------------------------------------------------------------------

export type DiscoveryAction = 'like' | 'pass' | 'save';

// ---------------------------------------------------------------------------
// Filter
// ---------------------------------------------------------------------------

export interface DiscoveryFilter {
  sport?: Sport;                 // required filter — gym, golf, tennis, or running
  level?: FitnessLevel;
  suburb?: string;               // Sydney suburb
  preferredTime?: PreferredTime;
  limit?: number;
  offset?: number;
}

// ---------------------------------------------------------------------------
// Response shapes
// ---------------------------------------------------------------------------

/**
 * A potential workout partner as shown in the discovery feed.
 * Intentionally limited — enough to render a card and make a like/pass decision.
 */
export interface PartnerCard {
  userId: UUID;
  displayName: string;
  suburb?: string;
  bioExcerpt?: string;    // truncated at 160 chars by the API
  avatarUrl?: string;
  age?: number;           // derived from birthYear on the server
  sportProfiles: Array<{
    sport: Sport;
    level: FitnessLevel;
    gymName?: string;
    golfClub?: string;
  }>;
}

export type DiscoveryFeedResponse = Paginated<PartnerCard>;

// ---------------------------------------------------------------------------
// Request / response for recording an action
// ---------------------------------------------------------------------------

export interface RecordActionRequest {
  targetUserId: UUID;
  action: DiscoveryAction;
  sport: Sport;
}

export interface RecordActionResponse {
  action: DiscoveryAction;
  matchCreated: boolean;
  matchId?: UUID;
}
